File: trading_api/go_analysis/handlers/patterns.go
package handlers

import (
	"go_analysis/models"
	"math"
	"net/http"

	"github.com/gin-gonic/gin"
)

// DetectPatterns handles POST /detect/patterns
func DetectPatterns(c *gin.Context) {
	var req models.PatternRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	ohlc := req.OHLC
	response := models.PatternResponse{
		Hammer:           make([]bool, len(ohlc)),
		InvertedHammer:   make([]bool, len(ohlc)),
		HangingMan:       make([]bool, len(ohlc)),
		DragonflyDoji:    make([]bool, len(ohlc)),
		GravestoneDoji:   make([]bool, len(ohlc)),
		BullishEngulfing: make([]bool, len(ohlc)),
		BearishEngulfing: make([]bool, len(ohlc)),
	}

	for i := 1; i < len(ohlc); i++ {
		candle := ohlc[i]
		body := math.Abs(candle.Close - candle.Open)
		totalRange := candle.High - candle.Low
		upperShadow := candle.High - math.Max(candle.Open, candle.Close)
		lowerShadow := math.Min(candle.Open, candle.Close) - candle.Low

		// Hammer: small body at top, long lower shadow
		if body < totalRange*0.3 && lowerShadow > body*2 && upperShadow < body*0.5 {
			response.Hammer[i] = true
		}

		// Inverted Hammer: small body at bottom, long upper shadow
		if body < totalRange*0.3 && upperShadow > body*2 && lowerShadow < body*0.5 {
			response.InvertedHammer[i] = true
		}

		// Hanging Man: same as hammer but in uptrend
		if body < totalRange*0.3 && lowerShadow > body*2 && upperShadow < body*0.5 {
			response.HangingMan[i] = true
		}

		// Dragonfly Doji: open/close at high, long lower shadow
		if body < totalRange*0.1 && lowerShadow > totalRange*0.6 {
			response.DragonflyDoji[i] = true
		}

		// Gravestone Doji: open/close at low, long upper shadow
		if body < totalRange*0.1 && upperShadow > totalRange*0.6 {
			response.GravestoneDoji[i] = true
		}

		// Bullish Engulfing
		if i > 0 {
			prev := ohlc[i-1]
			if prev.Close < prev.Open && candle.Close > candle.Open &&
				candle.Open < prev.Close && candle.Close > prev.Open {
				response.BullishEngulfing[i] = true
			}

			// Bearish Engulfing
			if prev.Close > prev.Open && candle.Close < candle.Open &&
				candle.Open > prev.Close && candle.Close < prev.Open {
				response.BearishEngulfing[i] = true
			}
		}
	}

	c.JSON(http.StatusOK, response)
}

File: trading_api/rust_analysis/src/patterns.rs
use crate::OHLC;

/// Detect Hammer pattern
pub fn detect_hammer(ohlc: &[OHLC]) -> Vec<bool> {
    let len = ohlc.len();
    let mut result = vec![false; len];
    
    for i in 0..len {
        let body = (ohlc[i].close - ohlc[i].open).abs();
        let upper_wick = ohlc[i].high - ohlc[i].close.max(ohlc[i].open);
        let lower_wick = ohlc[i].close.min(ohlc[i].open) - ohlc[i].low;
        let total_range = ohlc[i].high - ohlc[i].low;
        
        if total_range > 0.0 {
            // Hammer: small body, long lower wick, small upper wick
            if lower_wick > body * 2.0 && upper_wick < body * 0.5 {
                result[i] = true;
            }
        }
    }
    
    result
}

/// Detect Inverted Hammer pattern
pub fn detect_inverted_hammer(ohlc: &[OHLC]) -> Vec<bool> {
    let len = ohlc.len();
    let mut result = vec![false; len];
    
    for i in 0..len {
        let body = (ohlc[i].close - ohlc[i].open).abs();
        let upper_wick = ohlc[i].high - ohlc[i].close.max(ohlc[i].open);
        let lower_wick = ohlc[i].close.min(ohlc[i].open) - ohlc[i].low;
        let total_range = ohlc[i].high - ohlc[i].low;
        
        if total_range > 0.0 {
            // Inverted Hammer: small body, long upper wick, small lower wick
            if upper_wick > body * 2.0 && lower_wick < body * 0.5 {
                result[i] = true;
            }
        }
    }
    
    result
}

/// Detect Hanging Man pattern
pub fn detect_hanging_man(ohlc: &[OHLC]) -> Vec<bool> {
    // Hanging Man is same shape as Hammer but appears in uptrend
    // For now, use same logic as Hammer
    detect_hammer(ohlc)
}

/// Detect Bullish Engulfing pattern
pub fn detect_bullish_engulfing(ohlc: &[OHLC]) -> Vec<bool> {
    let len = ohlc.len();
    let mut result = vec![false; len];
    
    for i in 1..len {
        let prev_bearish = ohlc[i-1].close < ohlc[i-1].open;
        let curr_bullish = ohlc[i].close > ohlc[i].open;
        
        if prev_bearish && curr_bullish {
            // Current candle engulfs previous
            if ohlc[i].open < ohlc[i-1].close && ohlc[i].close > ohlc[i-1].open {
                result[i] = true;
            }
        }
    }
    
    result
}

/// Detect Bearish Engulfing pattern
pub fn detect_bearish_engulfing(ohlc: &[OHLC]) -> Vec<bool> {
    let len = ohlc.len();
    let mut result = vec![false; len];
    
    for i in 1..len {
        let prev_bullish = ohlc[i-1].close > ohlc[i-1].open;
        let curr_bearish = ohlc[i].close < ohlc[i].open;
        
        if prev_bullish && curr_bearish {
            // Current candle engulfs previous
            if ohlc[i].open > ohlc[i-1].close && ohlc[i].close < ohlc[i-1].open {
                result[i] = true;
            }
        }
    }
    
    result
}
